src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -235,6 +249,140 @@ public bool Read()
             return this.ReadInternal();
         }
 
+        /// <summary>
+        /// The value of the last reported node.
+        /// </summary>
+        /// <remarks>
+        /// Depending on whether buffering is on or off this will return the node value of the last
+        /// buffered read or the node value of the last unbuffered read.
+        /// </remarks>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the node value of the last buffered read if buffering is on or off;
+        /// otherwise the node value of the last unbuffered read.</returns>
+        public Task<object> GetValueAsync()
+        {
+            this.AssertAsynchronous();
+
+            if (this.bufferedNodesHead != null)
+            {
+                if (this.isBuffering)
+                {
+                    Debug.Assert(this.currentBufferedNode != null, "this.currentBufferedNode != null");
+                    return Task.FromResult(this.currentBufferedNode.Value);
+                }
+
+                // in non-buffering mode if we have buffered nodes satisfy the request from the first node there
+                return Task.FromResult(this.bufferedNodesHead.Value);
+            }
+
+            return this.asyncInnerReader.GetValueAsync();
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the input.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains true if a new node was found, or false if end of input was reached.</returns>
+        public Task<bool> ReadAsync()
+        {
+            Debug.Assert(!this.parsingInStreamError, "!this.parsingInStreamError");
+            this.AssertAsynchronous();
+
+            // read the next node and check whether it is an in-stream error
+            return this.ReadInternalAsync();
+        }
+
+        /// <summary>
+        /// Asynchronously checks whether or not the current value can be streamed.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous operation.
+        /// true if the current value can be streamed; otherwise false.</returns>
+        public virtual async Task<bool> CanStreamAsync()
+        {
+            this.AssertAsynchronous();
+
+            if (!this.isBuffering && this.asyncInnerReader is IJsonStreamReaderAsync asyncStreamReader)

```

Comment:
I see this check for `IJsonStreamReaderAsync ` in a few different places. I wonder if it makes sense to have a field where this is stored so that the casting only has to happen once. I also wonder if that field should be set in the constructor. I can see there being different use cases where it makes sense to do it on the fly versus doing it in the constructor. Any thoughts?